File: course2/week1/1_brackets.js
'use strict';

var open = { '[': ']', '{': '}', '(': ')' };
var close = { ']': '[', '}': '{', ')': '(' };

function check(string) {
    var stack = [];
    for (var i = 0; i < string.length; i++) {
        var char = string[i];
        if (open[char]) {
            stack.push({ char: char, position: i });
        }
        else if (close[char]) {
            var match = stack.pop();
            if (!match || match.char !== close[char]) return i + 1
        }
    }
    if (!stack.length) return 'Success';
    return stack.pop().position + 1;
}

function fromStdin() {
    var rl = require('readline').createInterface({ input: process.stdin, output: process.stdout });
    rl.on('line', function(line) {
        console.log(check(line));
        rl.close();
    });
}
fromStdin();

// function test(string, correct) {
//     var result = check(string);
//     console.log(result.toString() == correct.toString() ? 'ok' : `wrong: ${result} instead of ${correct}`);
// }
//
// test('[]', 'Success');
// test('{}[]', 'Success');
// test('[{}]', 'Success');
// test('(())', 'Success');
// test('{aaaaa', 1);
// test('{[}', 3);
// test('foo(bar);', 'Success');
// test('foo(bar[i);', 10);

File: course2/week1/1_brackets.py
# python2

import sys

stack = []
open = '([{'
close = {']': '[', '}': '{', ')': '('}


def check(string):
    for index, char in enumerate(string):
        if char in open:
            stack.append((char, index))
        elif char in close:
            if not stack:
                return index + 1
            match = stack.pop()
            if not close[char] == match[0]:
                return index + 1
    if stack:
        return stack.pop()[1] + 1
    else:
        return 'Success'


text = sys.stdin.readline()
print(check(text))

# def test(input, expected):
#     res = check(input)
#     print 'ok' if res == expected else 'wrong: %s insted of %s' % (res, expected)

# test('[]', 'Success')
# test('{}[]', 'Success')
# test('[{}]', 'Success')
# test('(())', 'Success')
# test('{aaaaa', 1)
# test('{[}', 3)
# test('foo(bar);', 'Success')
# test('foo(bar[i);', 10)
